src/conn.c

```diff
@@ -245,6 +262,56 @@ void xmpp_conn_set_keepalive(xmpp_conn_t *conn, int timeout, int interval)
     }
 }
 
+/** Register sockopt callback
+ *  Set function to be called when a new socket is created to allow setting
+ *  socket options before connection is started.
+ *
+ *  @param conn The Strophe connection object this callback is being registered
+ * for
+ *  @param callback a xmpp_sockopt_callback callback function that will receive
+ *      notifications of connection status
+ *
+ *  @ingroup Connections
+ */
+
+void xmpp_conn_set_sockopt_callback(xmpp_conn_t *conn,
+                                    xmpp_sockopt_callback callback)
+{
+    strophe_debug(conn->ctx, "xmpp", "Registering sockopt callback");
+
+    if (conn->ka_timeout || conn->ka_interval)
+        strophe_warn(
+            conn->ctx, "xmpp",
+            "sockopt callback registered after xmpp_conn_set_keepalive(). "
+            "xmpp_conn_set_keepalive() is ignored.");
+
+    conn->sockopt_cb = callback;
+}
+
+/** Example sockopt callback function
+ *  An example function that can be used to set reasonable default keepalive
+ * options on sockets when registered for a connection with
+ * xmpp_conn_set_sockopt_callback()
+ *
+ *  @ingroup Connections
+ */
+int xmpp_sockopt_cb_keepalive(xmpp_conn_t *conn, void *socket)
```

Comment:
I very much like! :+1: 

Can you please move this to `sock.c`? like that you can also remove the `ifdef` and use `sock_t sock = *((sock_t *)socket);` instead

Reply:
I didn't do this originally because `sock.c` was independent from rest of libstrophe, but since I then changed that to call the callback function from sock.c anyway ...